include/kernels/gauge_shift.cuh

```diff
@@ -0,0 +1,78 @@
+#pragma once
+
+#include <gauge_field_order.h>
+#include <quda_matrix.h>
+#include <index_helper.cuh>
+#include <kernel.h>
+
+namespace quda {
+
+  template <typename Float_, int nColor_, QudaReconstructType recon_u>
+  struct GaugeShiftArg : kernel_param<> {
+    using Float = Float_;
+    static constexpr int nColor = nColor_;
+    static_assert(nColor == 3, "Only nColor=3 enabled at this time");
+    typedef typename gauge_mapper<Float,recon_u>::type Gauge;
+
+    Gauge out;
+    const Gauge in;
+    int geometry;
+
+    int S[4]; // the regular volume parameters
+    int X[4]; // the regular volume parameters
+    int E[4]; // the extended volume parameters
+    int border[4]; // radius of border
+    int P; // change of parity
+
+    GaugeShiftArg(GaugeField &out, const GaugeField &in, const int* dx) :
+      kernel_param(dim3(in.VolumeCB(), 2, in.Geometry())),
+      out(out),
+      in(in),
+      geometry(in.Geometry())
+    {
+      P = 0;
+      for (int i=0; i<4; i++) {
+	S[i] = dx[i];
+	X[i] = out.X()[i];
+	E[i] = in.X()[i];
+	border[i] = (E[i] - X[i])/2;
+	P += dx[i];
+      }
+      P = std::abs(P)%2;
+    }
+  };
+
+  template <typename Arg, int dir>
```

Comment:
There's no need here for `dir` to be a template.  It's not being used to index into a thread-local array, so making it a template only really adds bloat.

Reply:
https://github.com/lattice/quda/pull/1348/commits/666a38e49176dbd6e365d42b1e45872fb76b25c9